File: csharp/edit_text/edit_text.cs
// ============================================================================
//  edit_text -- C# port of examples\Vb6\edit_text\edit_text.bas
//  Imports a PDF, then uses the content parser to find and replace every
//  occurrence of a search string on each page.
// ============================================================================
using System;
using System.IO;
using System.Runtime.InteropServices;
using LumasPdfSdk;

class EditText
{
    static TErrorProc _err = ErrProc;

    static int ErrProc(IntPtr data, int errCode, string errMessage, int errType)
    {
        Console.WriteLine(errMessage);
        return 0;                        // try to continue on error
    }

    static void Main()
    {
        string exeDir = AppDomain.CurrentDomain.BaseDirectory;

        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfCreateNewPDFW(pdf, "");            // output file opened later
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);
        // Avoid the conversion of pages to templates
        LumasPdf.pdfSetImportFlags(pdf, LumasPdfConsts.ifImportAll | LumasPdfConsts.ifImportAsPage);

        string inFile = Path.Combine(exeDir, "sample_multipage.pdf");
        if (LumasPdf.pdfOpenImportFileW(pdf, inFile, (int)LumasPdfConsts.ptOpen, "") < 0)
        {
            LumasPdf.pdfDeletePDF(pdf);
            return;
        }
        LumasPdf.pdfImportPDFFile(pdf, 1, 1.0, 1.0);
        LumasPdf.pdfCloseImportFile(pdf);

        IntPtr ctx = LumasPdf.psrCreateParserContext(pdf, LumasPdfConsts.ofDefault, IntPtr.Zero);
        string searchText = "PDF";                     // occurs very often in the help file
        string replaceText = "XDF";                    // just an example

        IntPtr selPtr = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(TTextSelection)));
        try
        {
            for (int i = 1; i <= LumasPdf.pdfGetPageCount(pdf); i++)
            {
                TContent content = new TContent();
                // The flag cpfEnableTextSelection is required, otherwise no text can be found.
                if (LumasPdf.psrParsePage(pdf, ctx, IntPtr.Zero, IntPtr.Zero, (uint)i,
                        LumasPdfConsts.cpfEnableTextSelection, IntPtr.Zero, ref content))
                {
                    IntPtr curr = IntPtr.Zero;
                    TTextSelection sel = new TTextSelection();
                    while (LumasPdf.psrFindText(pdf, ctx, IntPtr.Zero, (uint)LumasPdfConsts.stDefault, curr,
                                searchText, (uint)searchText.Length, ref sel))
                    {
                        LumasPdf.psrReplaceSelText(pdf, ctx, TReplaceTextFlags.rtfDefault, ref sel,
                            replaceText, (uint)replaceText.Length);
                        Marshal.StructureToPtr(sel, selPtr, false);
                        curr = selPtr;
                    }
                    LumasPdf.psrWriteToPage(pdf, ctx, LumasPdfConsts.ofDefault, IntPtr.Zero);
                }
            }
        }
        finally
        {
            Marshal.FreeHGlobal(selPtr);
        }
        LumasPdf.psrDeleteParserContext(ref ctx);

        string outFile = Path.Combine(exeDir, "out.pdf");
        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile))
            {
                LumasPdf.pdfDeletePDF(pdf);
                return;
            }
        }
        if (LumasPdf.pdfCloseFile(pdf))
            Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");
        LumasPdf.pdfDeletePDF(pdf);
    }
}

File: cpp/edit_text/edit_text.cpp
// edit_text -- C++ port of examples\Vb6\edit_text\edit_text.bas
// Imports a PDF and uses the content parser to find/replace a string on each page.
#include <lumaspdf.h>
#include "repo_root.h"
#include <cstdio>
#include <cstring>
#include <string>

static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    if(ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

int main(int argc, char** argv){
    PPDF pdf = pdfNewPDF();
    pdfCreateNewPDFA(pdf, "");
    pdfSetOnErrorProc(pdf, nullptr, ErrProc);
    pdfSetImportFlags(pdf, ifImportAll | ifImportAsPage);

    const char* inFile = LUMAS_REPO_ROOT "/dynapdf_help.pdf";
    if(pdfOpenImportFileA(pdf, inFile, ptOpen, "") < 0){ pdfDeletePDF(pdf); return 1; }
    pdfImportPDFFile(pdf, 1, 1.0, 1.0);
    pdfCloseImportFile(pdf);

    IPSR ctx = psrCreateParserContext(pdf, ofDefault, nullptr);
    const wchar_t* searchText = L"PDF";
    const wchar_t* replaceText = L"XDF";
    UI32 searchLen = (UI32)wcslen(searchText);
    UI32 replaceLen = (UI32)wcslen(replaceText);

    TContent content{};
    TTextSelection sel{};
    for(SI32 i = 1; i <= pdfGetPageCount(pdf); ++i){
        if(psrParsePage(pdf, ctx, nullptr, nullptr, i, cpfEnableTextSelection, nullptr, &content) != 0){
            PTextSelection last = nullptr;
            while(psrFindText(pdf, ctx, nullptr, stDefault, last, (LWCHAR*)searchText, searchLen, &sel) != 0){
                psrReplaceSelText(pdf, ctx, rtfDefault, &sel, (LWCHAR*)replaceText, replaceLen);
                last = &sel;
            }
            psrWriteToPage(pdf, ctx, ofDefault, nullptr);
        }
    }
    psrDeleteParserContext(&ctx);

    std::string outFile = exeDir(argv[0]) + "/out.pdf";
    if(pdfHaveOpenDoc(pdf) != 0){
        if(pdfOpenOutputFileA(pdf, outFile.c_str()) == 0){ pdfDeletePDF(pdf); return 1; }
    }
    if(pdfCloseFile(pdf) != 0) printf("PDF file \"%s\" successfully created!\n", outFile.c_str());
    pdfDeletePDF(pdf);
    return 0;
}
